functions-1/add-prefix-suffix-ngram-udfs.md

```diff
@@ -0,0 +1,141 @@
+
+## What is N-gram
+
+N-gram breaks down text into overlapping sequences of N characters. For example:
+
+**"Apache Pinot"** with 2-grams produces:
+- `Ap`, `pa`, `ac`, `ch`, `he`, `e `, ` P`, `Pi`, `in`, `no`, `ot`
+
+When matching `ngram(col, "ap")`, it searches for the substring "ap" within the generated n-grams.
+When matching `ngram(col, 'apache')`, it generates n-grams from "apache" (ap, pa, ac, ch, he) and matches against the column's n-grams.
+
+## When to Use N-gram
+
+N-grams are particularly useful for:
+
+- **Text matching** - Fast substring search
+- **Wildcard queries** - Alternative to `REGEXP_LIKE` (which requires full scan)
+- **Prefix matching** - More efficient than O(N * length of string) operations
+- **Prefiltering** - Avoid full table scans by using n-gram indexing
+
+### Performance Comparison
+- `REGEXP_LIKE`: Full scan O(N)
+- `Prefix matching`: O(N * length of string)
+- `Wildcard matching`: O(N * length)
+- **N-gram approach**: Enables prefiltering to avoid scanning all rows
+
+## Signature
+
+> NGRAM(col, n)
+>
+> PREFIX(col, prefixString)
+>
+> SUFFIX(col, suffixString)
+
+### `col`
+The input string column or expression to process.
+
+### `n`
+For n-gram operations, the length of each gram (typically 2 for bigrams, 3 for trigrams).
+
+### `prefixString` / `suffixString`
+The string to add as prefix or suffix to the input.
+
+## Usage Examples
+
+### N-gram Generation
+
+```sql
+SELECT NGRAM('Apache Pinot', 2) AS bigrams
+FROM myTable
+```
+
+| bigrams |
+| ------- |
+| Ap,pa,ac,ch,he,e ,  P,Pi,in,no,ot |
+
+```sql
+SELECT NGRAM('fast', 3) AS trigrams
+FROM myTable
+```
+
+| trigrams |
+| -------- |
+| fas,ast |
+
+### Prefix Operations
+
+```sql
+SELECT PREFIX('Pinot', 'Apache ') AS name
+FROM myTable
+```
+
+| name         |
+| ------------ |
+| Apache Pinot |
+
+### Suffix Operations
+
+```sql
+SELECT SUFFIX('Apache', ' Pinot') AS name
+FROM myTable
+```
+
+| name         |
+| ------------ |
+| Apache Pinot |
+
+### Advanced Text Matching with N-grams
+
+Using n-grams for efficient text search with prefiltering:
+
+```sql
+SELECT organizationUUID, confirmedEmployeeCount, name
+FROM rta.rta.u4b_organizations
+WHERE deletedAt IS NULL
+  AND (IN_SERIALIZED_LIST(entityTypes, 'org!https://emoji.slack-edge.com/TQ3BASMK9/u4b/b32791585b8da8cc.jpg!organization'))
+  AND (TEXT_MATCH(name, '/.pacific./ AND /.equity./')
+       OR TEXT_MATCH(profileName, '/.pacific./ AND /.equity./'))
+ORDER BY confirmedEmployeeCount DESC
+OFFSET 0 LIMIT 10
+```
+
+### N-gram Based Wildcard Search
+
+```sql
+-- Traditional approach (slow - full scan)
+SELECT * FROM documents
+WHERE REGEXP_LIKE(content, '.*apache.*pinot.*')
+
+-- N-gram approach (fast - with prefiltering)
+SELECT * FROM documents
+WHERE NGRAM(content, 2) LIKE '%ap%'
```

Comment:
Why would we use LIKE here? We can simply do `NGRAM(content, 2) = 'ap'`?

Reply:
thank you Ankit, would address this